### DISCUSSION TITLE
How to connect SNodes only by SPorts

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#921] [Q&A] [answered]
Hi,
    I create a SNode and some SPorts in it. I only want to connect SNodes by SPorts , but I find when I click tool palatte's edge button and move over SNode I can click and connect them. I look for sprotty code and find snode has a canConnect method. 
 ![image](https://user-images.githubusercontent.com/4537809/218688993-785a9e84-9ca4-4e6e-b1d3-8c04d4e86de5.png)
 SNode's default canConnect is what I want. but there is an ApplyTypeHintsCommand class overrides SNode's default canConnect. I do not know where invoke the ApplyTypeHintsCommand.  and how to only connect SNodes only by SPorts
![image](https://user-images.githubusercontent.com/4537809/218689083-33e803f1-39e4-4d94-8213-5eee4b332469.png)

**Comment by tortmayr** [comment, marked as answer]
Hi @wss29,
in GLSP such constraints are tyically configured via `TypeHints` on the server side.
This is done in the [diagram configuration](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/node-json-theia/glsp-server/src/diagram/tasklist-diagram-configuration.ts). Here you can specify the typehints for specific model types.
If you want to only allow ports as connection source/target you have to provide/adapt the typehint for your edge and only allow the type or your ports e.g.:

```ts
 get edgeTypeHints(): EdgeTypeHint[] {
        return [
            {
                elementTypeId: DefaultTypes.EDGE,
                deletable: true,
                repositionable: false,
                routable: false,
                sourceElementTypeIds: [DefaultTypes.PORT],
                targetElementTypeIds: [DefaultTypes.PORT]
            }
        ];
```

**Comment by wss29** [comment]
Thank a great, this save me a lot of time
